File: Java/interview_29.java
class Solution {
    public int[] spiralOrder(int[][] matrix) {
        int row = matrix.length;
        if (row == 0) {
            return new int[0];
        }
        int col = matrix[0].length;
        int totalCount = row * col;
        int[] result = new int[totalCount];
        if (totalCount == 0) {
            return result;
        }
        int left = 0, right = col - 1, top = 0, bottom = row - 1;
        int idx = 0;
        while (idx < totalCount) {
            for (int j = left; j <= right; ++j) {
                result[idx++] = matrix[top][j];
            }
            ++top;
            if (idx >= totalCount) {
                break;
            }
            for (int i = top; i <= bottom; ++i) {
                result[idx++] = matrix[i][right];
            }
            --right;
            if (idx >= totalCount) {
                break;
            }
            for (int j = right; j >= left; --j) {
                result[idx++] = matrix[bottom][j];
            }
            --bottom;
            if (idx >= totalCount) {
                break;
            }
            for (int i = bottom; i >= top; --i) {
                result[idx++] = matrix[i][left];
            }
            ++left;
        }
        return result;
    }
}

File: python/interview_29.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        row = len(matrix)
        if row == 0:
            return []
        col = len(matrix[0])
        if col == 0:
            return []
        result = []
        top, bottom = 0, row - 1
        left, right = 0, len(matrix[0]) - 1
        count = row * col
        idx = 0
        while idx < count:
            for j in range(left, right + 1):
                result.append(matrix[top][j])
                idx += 1
            if idx >= count:
                break
            top += 1
            for i in range(top, bottom + 1):
                result.append(matrix[i][right])
                idx += 1
            if idx >= count:
                break
            right -= 1
            for j in range(right, left - 1, -1):
                result.append(matrix[bottom][j])
                idx += 1
            if idx >= count:
                break
            bottom -= 1
            for i in range(bottom, top - 1, -1):
                result.append(matrix[i][left])
                idx += 1
            left += 1
        return result
        
